lib/ddtrace/profiling/tasks/setup.rb

```diff
@@ -59,32 +44,60 @@ def activate_cpu_extensions
           log "[DDTRACE] CPU profiling unavailable. Cause: #{e.message} Location: #{e.backtrace.first}"
         end
 
-        def autostart_profiler
-          if Datadog::Profiling.supported?
-            # Start the profiler
-            Datadog.profiler.start if Datadog.profiler
+        def setup_at_fork_hooks
+          if Process.respond_to?(:at_fork)
+            Process.at_fork(:child) do
+              begin
+                # When Ruby forks, clock IDs for each of the threads
+                # will change. We can only update these IDs from the
+                # execution context of the thread that owns it.
+                # This hook will update the IDs for the main thread
+                # after a fork occurs.
+                if Thread.current.respond_to?(:update_native_ids, true)
+                  Thread.current.send(:update_native_ids)
+                end
 
-            # Setup at_fork hook:
-            # When Ruby forks, threads running in the parent process
-            # won't be restarted in the child process. This hook will
-            # restart the profiler in the child process when this happens.
-            if Process.respond_to?(:at_fork)
-              Process.at_fork(:child) { Datadog.profiler.start if Datadog.profiler }
+                # Restart profiler, if enabled
+                Datadog.profiler.start if Datadog.profiler
+              rescue StandardError => e
+                log "[DDTRACE] Error during post-fork hooks. Cause: #{e.message} Location: #{e.backtrace.first}"
+              end
             end
-          elsif Datadog.configuration.profiling.enabled
-            # Log warning if profiling was supposed to be activated.
-            log '[DDTRACE] Profiling did not autostart; profiling not supported.'
           end
-        rescue StandardError => e
-          log "[DDTRACE] Could not autostart profiling. Cause: #{e.message} Location: #{e.backtrace.first}"
         end
 
-        private
-
         def log(message)
           # Print to STDOUT for now because logging may not be setup yet...
           puts message
         end
+
+        # Small helper class to allow some piece of code to be run only once
+        class OnlyOnce
```

Comment:
We have something similar in our integration code: https://github.com/DataDog/dd-trace-rb/blob/079f18f410f590252030578aa240576f3fc4c6e5/lib/ddtrace/patcher.rb#L24-L40

It doesn't do the same thing, has different tracking features, and it's not thread safe (as it doesn't need to be).

I don't see these two implementations easily merging, but I think that `class OnlyOnce` is more generic and simple than `def do_once`. I think it might make sense to move `OnlyOnce` to the global tracer `lib` namespace (outside of profiling) so it can be used by the tracer as well.

Ideally we could even implement `def do_once` on top of `OnlyOnce` in the future.

Reply:
I've taken a stab at this -- @marcotc let me know what you think of https://github.com/DataDog/dd-trace-rb/pull/1398 .

Reply:
Looks great, left a comment in that PR, but the implementation seems like it works great.

Reply:
Would you mind if I merged this change and then later when that one is complete I'll just remove the inlined version and replace it with the common `OnlyOnce` implementation? It seems like there's still some concerns with that one so I'd prefer to decouple the work.

I am asking because I know this PR only targets the profiling branch (so I could just say "merge all the things"), but I prefer to treat it as if it was master, because I'm expecting that pretty soon it will be merged to master so I want to keep the code peer reviewed and the quality high :)